### PR TITLE
Implements HTTP GET node endpoint

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**: Implements HTTP GET node endpoint.

**Special notes for your reviewer**: https://github.com/kubermatic/kubermatic/issues/1371

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: api/pkg/handler/routes_v1.go [hunk 1]
```diff
@@ -811,3 +817,31 @@ func (r Routing) detachSSHKeyFromCluster() http.Handler {
...
+//
+//     Consumes:
+//     - application/json
+//
+//     Produces:
+//     - application/json
+//
+//     Responses:
+//       default: errorResponse
+//       200: NodeV2
```
Comment (p0lyn0mial) [inline api/pkg/handler/routes_v1.go]: @kgroschoff @mrIncompetent do you think we could slim down the `Node` resource that is returned to the client ?

Here is the definition of the resource:
https://github.com/kubermatic/kubermatic/blob/master/api/pkg/api/v2/node.go#L9

Comment (mrIncompetent) [reply api/pkg/handler/routes_v1.go]: I think we use almost all of this :thinking:

Comment (kgroschoff) [reply api/pkg/handler/routes_v1.go]: What @mrIncompetent said.

Comment (p0lyn0mial) [reply api/pkg/handler/routes_v1.go]: sgtm, thanks!

File: api/pkg/api/v1/types.go [hunk 2]
```diff
@@ -282,3 +274,19 @@ type NewClusterHealth struct {
 // NewClusterList represents a list of clusters
 // swagger:model ClusterList
 type NewClusterList []NewCluster
+
+// Node represents a worker node that is part of a cluster
+// swagger:model Node
+type Node struct {
+	NewObjectMeta `json:",inline"`
+
+	// TODO: normally referring to a field that is defined in v2 is bad, if you are doing this please stop
```
Comment (p0lyn0mial) [inline api/pkg/api/v1/types.go:301]: what do you think ?

File: api/pkg/handler/node_v1.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		if err != nil {
+			return nil, kubernetesErrorToHTTPError(err)
+		}
+
+		cluster, err := clusterProvider.Get(user, project, req.ClusterName)
+		if err != nil {
+			return nil, kubernetesErrorToHTTPError(err)
+		}
+
+		machineClient, err := clusterProvider.GetMachineClientForCustomerCluster(cluster)
```
Comment (p0lyn0mial) [inline api/pkg/handler/node_v1.go:41]: in general there is no special bond (`OwnerReferences` or `Labels`) between the cluster and the nodes. We just need to open the connection to the user's cluster and get/create nodes/machines.

File: api/pkg/handler/node_v1.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		if err != nil {
+			return nil, kubernetesErrorToHTTPError(err)
+		}
+
+		kubeClient, err := clusterProvider.GetKubernetesClientForCustomerCluster(cluster)
+		if err != nil {
+			return nil, kubernetesErrorToHTTPError(err)
+		}
+
+		machine, node, err := tryToFindMachineAndNode(req.NodeName, machineClient, kubeClient)
```
Comment (p0lyn0mial) [inline api/pkg/handler/node_v1.go:56]: normally we have `project`, `user` and `sshkey` providers but here we decided to use `machineClient` and `kubeClient` directly to access the user cluster. 

how about moving `machineClient` and `kubeClient` to their own provider ?

File: api/pkg/api/v1/types.go (hunk 2 above)
Comment (mrIncompetent) [reply api/pkg/api/v1/types.go:301]: sgtm

File: api/pkg/handler/node_v1.go (hunk 3 above)
Comment (mrIncompetent) [reply api/pkg/handler/node_v1.go:41]: sgtm

File: api/pkg/handler/node_v1.go (hunk 4 above)
Comment (mrIncompetent) [reply api/pkg/handler/node_v1.go:56]: I would keep it that way until we have a good solution on machine+node management.
I'm working on a proposal which would eventually change the way we handle machines/nodes in user-clusters

File: api/pkg/handler/node_v1.go [hunk 5]
```diff
@@ -0,0 +1,111 @@
+package handler
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/go-kit/kit/endpoint"
+
+	"fmt"
```
Comment (mrIncompetent) [inline api/pkg/handler/node_v1.go]: that little evil ;)

File: api/pkg/handler/node_v1_test.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	apiv1 "github.com/kubermatic/kubermatic/api/pkg/api/v1"
+	kubermaticv1 "github.com/kubermatic/kubermatic/api/pkg/crd/kubermatic/v1"
+	"github.com/kubermatic/kubermatic/api/pkg/kubernetes"
+	"github.com/kubermatic/machine-controller/pkg/machines/v1alpha1"
+
+	"fmt"
```
Comment (mrIncompetent) [inline api/pkg/handler/node_v1_test.go]: that little evil ;)

File: api/pkg/handler/routing_test.go [hunk 7]
```diff
@@ -195,3 +203,20 @@ func TestUpRoute(t *testing.T) {
 	ep.ServeHTTP(res, req)
 	checkStatusCode(http.StatusOK, res, t)
 }
+
+type fakeUserClusterConnection struct {
```
Comment (mrIncompetent) [inline api/pkg/handler/routing_test.go:219]: love it!

File: api/pkg/handler/node_v1.go (hunk 5 above)
Comment (p0lyn0mial) [reply api/pkg/handler/node_v1.go]: that's embarrassing :( :P

File: api/pkg/handler/node_v1_test.go (hunk 6 above)
Comment (p0lyn0mial) [reply api/pkg/handler/node_v1_test.go]: no comment.

**Comment by p0lyn0mial** [comment]
ready for final review.

**Review by mrIncompetent — APPROVED** [review]

**Pull request merged**